tests/functional/conftest.py

```diff
@@ -1,100 +1,354 @@
+
+
+@when(
+    parsers.cfparse(
+        'the query parameters contain {key:String} as {value:String}',
+        extra_types=dict(String=str)
+    ),
+    target_fixture='query_params',
+)
+def amended_query_params(search: Search, key: str, value: str) -> str:
+    query_params = copy(search.query)
+    query_params.append((key, value))
+    return urllib.parse.urlencode(query_params)
+
+
+@when("I search for the patient's PDS record", target_fixture='response')
+def search_patient(headers_with_authorization: dict, query_params: str, pds_url: str) -> Response:
+    return get(url=f"{pds_url}/Patient?{query_params}", headers=headers_with_authorization)
+
+
+@then(
+    parsers.cfparse(
+        "I get a {expected_status:Number} HTTP response code",
+        extra_types=dict(Number=int)
+    )
+)
+def check_status(response: Response, expected_status: int) -> None:
+    with check:
+        assert response.status_code == expected_status
+
+
+@then(
+    parsers.cfparse(
+        'the response body is the {error:String} response',
+        extra_types=dict(String=str)
+    )
+)
+def response_body_contains_error(response_body: dict, error) -> None:
+    assert response_body == RESPONSES[error]
+
+
+@then('the response body contains the expected response')
+def response_body_as_expected(response_body: dict, patient: Patient) -> None:
+    assert response_body == patient.expected_response
+
+
+@then(
+    parsers.cfparse(
+        'the {header_field:String} response header matches the request',
+        extra_types=dict(String=str)
+    )
+)
+def check_header_value(response: Response,
+                       header_field: str,
+                       headers_with_authorization: dict) -> None:
+    with check:
+        assert response.headers[header_field] == headers_with_authorization[header_field]
+
+
+@then("the response body contains the patient's NHS number")
+def response_body_contains_given_id(response_body: dict, nhs_number: dict) -> None:
+    with check:
+        assert response_body["id"] == nhs_number
+        assert response_body["resourceType"] == "Patient"
+
+
+@then('the response body is the correct shape')
+def response_body_shape(response_body: Response) -> None:
+    with check:
+        assert response_body["address"] is not None
+        assert isinstance(response_body["address"], list)
+
+        assert response_body["birthDate"] is not None
+        assert isinstance(response_body["birthDate"], str)
+        assert len(response_body["birthDate"]) > 1
+
+        assert response_body["gender"] is not None
+        assert isinstance(response_body["gender"], str)
+
+        assert response_body["name"] is not None
+        assert isinstance(response_body["name"], list)
+        assert len(response_body["name"]) > 0
+
+        assert len(response_body["identifier"]) > 0
+        assert isinstance(response_body["identifier"], list)
+
+        assert response_body["meta"] is not None
```

Comment:
Not a huge fan of this `then` step. But left it here to keep the `test_user_restricted` tests as close to their original checks as possible.